tests/plt/plot_test.py

```diff
@@ -0,0 +1,189 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @file
+
+import scipp as sc
+from scipp.plt import Plot, Model, Figure, widgets, Node
+from ..factory import make_dense_data_array, make_dense_dataset
+import matplotlib.pyplot as plt
+import matplotlib
+import ipywidgets as ipw
+
+matplotlib.use('Agg')
+
+
+def test_plot_single_1d_line():
+    da = make_dense_data_array(ndim=1)
+    m = Model(da)
+    fig = Figure()
+    m.add_view(m.root.name, fig)
+    p = Plot()
+    p.add_model("da", m)
+    p.render()
+
+
+def test_plot_two_1d_lines():
+    ds = make_dense_dataset(ndim=1)
+    m_a = Model(ds['a'])
+    m_b = Model(ds['b'])
+    fig = Figure()
+    m_a.add_view(m_a.root.name, fig)
+    m_b.add_view(m_b.root.name, fig)
+    p = Plot()
+    p.add_model("a", m_a)
+    p.add_model("b", m_b)
+    p.render()
+
+
+def test_plot_2d_image():
+    da = make_dense_data_array(ndim=2)
+    m = Model(da)
+    fig = Figure()
+    m.add_view(m.root.name, fig)
+    p = Plot()
+    p.add_model("da", m)
+    p.render()
+
+
+def test_plot_2d_image_smoothing_slider():
+    da = make_dense_data_array(ndim=2)
+    m = Model(da)
+    from scipy.ndimage import gaussian_filter
+
+    def smooth(da, sigma):
+        out = da.copy()
+        smoothed = sc.array(dims=da.dims,
+                            values=gaussian_filter(da.values, sigma=sigma),
+                            unit=da.unit)
+        out.data = smoothed
+        return out
+
+    sl = ipw.IntSlider(min=1, max=10)
+    smooth_node = Node(func=smooth)
+    smooth_view = widgets.WidgetView(widgets={"sigma": sl})
+    m.add("smooth", smooth_node, after=m.end.name)
+    f = Figure()
+    m.add_view("smooth", f)
+    m.add_view("smooth", smooth_view)
+    p = Plot()
+    p.add_model("da", m)
+    p.render()
+    sl.value = 5
+
+
+def test_plot_2d_image_with_masks():
+    da = make_dense_data_array(ndim=2)
+    da.masks['m1'] = da.data < sc.scalar(0.0, unit='counts')
+    da.masks['m2'] = da.coords['xx'] > sc.scalar(30., unit='m')
+    masks_node = Node(func=widgets.hide_masks)
+    masks_widget = widgets.MaskWidget(masks=da.masks)
+    masks_view = widgets.WidgetView(widgets={"masks": masks_widget})
+    m = Model(da)
+    m.add("hiding_masks", masks_node, after=m.end.name)
```

Comment:
Coming back to API design again:

What is the purpose of the string key (first argument)? Looking at all the examples here we are basically:
1. Make up some name and drop it into the model (with a new node).
2. Pull the name out again for the purpose of using it as the `after` for the next node.

Would it not feel more natural to write code like this (not suggesting this concretely, just an example)?
```python
m.add(node1, parent=node0)
m.add(node2, parent=node1)
```

It may be worth looking into how dask specifies and defines graphs. The [specification](https://docs.dask.org/en/stable/spec.html#definitions) looks similar to what we are doing here, which each string showing up as a key and as an "argument" to a follow-up task:
```python
{'x': 1,
 'y': 2,
 'z': (add, 'x', 'y'),
 'w': (sum, ['x', 'y', 'z']),
 'v': [(sum, ['w', 'z']), 2]}
```
but as a user we do not have to spell this key by hand (usually). Instead the [delayed](https://docs.dask.org/en/stable/delayed.html) interface can be used. Can we think of an equivalent approach for our interface?

```python
@scipp.plt.node
def func1():
    ...
@scipp.plt.node
def func2():
    ...

a = func1(model)
b = func2(a)
f = Figure(b)
p.append(f)  # or string key, if we need a key
```

Reply:
```Py
a = func1(model)
b = func2(a)
f = Figure(b)
p.append(f)  # or string key, if we need a key
```
may be a good approach, but I don't really get the final part where you append the Figure to the plot. Why add only the Figure, is it because it is only that which needs to be displayed?

In addition, I liked the use of string keys as entry points into a graph, because we might have base plotting functions, e.g. a 2d image with a slider through a 3d cube, that users could use to then add more views and nodes for customization.
If you don't have references to the node objects themselves, it's harder to attach new things to them?

Reply:
> may be a good approach, but I don't really get the final part where you append the Figure to the plot. Why add only the Figure, is it because it is only that which needs to be displayed?

Look at how dask graphs work, or compute outputs. You usually hold on to the tip(s) of the graph, and those have parents and grandparents. You don't work with the "root" of the graph once you have extended it downwards (unless you need to add more things attached to the root).

Reply:
> we might have base plotting functions, e.g. a 2d image with a slider through a 3d cube, that users could use to then add more views and nodes for customization.
If you don't have references to the node objects themselves, it's harder to attach new things to them?

Not sure how the strings make this easier? How do you know the string if a base-plotting function built the base graph?

Reply:
I think one of the key things that make it difficult to use the syntax you propose, is that the nodes in the graph only know about the one node they 'depend' on (their parent node). They do not know about their children. The graph is the one that knows this.

So when a node has to notify its children that it's about to change and they should update their views, the graph has to do that. I wanted to avoid each node holding a reference to the graph, because it should only need to know about the Views that are attached to it, and their parent Node.

The way it's currently done when a view need to send a notification down the graph is by passing it a method of the graph, with the `key` set using `partial`. It is not calling something on the node but on the graph.
```Py
def add_view(self, node, view):
        node.add_view(view)
        view.add_graph_node(node)
        view.add_notification(partial(self.notify_from_dependents, name=node.name))
```

If we change it so that each node can send notifications down the graph, then we could probably implement the approach you suggest